solarkraft/src/state/value.ts

```diff
@@ -16,7 +16,100 @@ export type Value = {
     | MapValue
 )
 
-export function isValid(v: Value): boolean {
+const EMPTY_COLLECTION_TYPE: string = 'T'
+const UNKNOWN_TYPE: string = 'UNKNOWN'
+
+// Helper function, takes an array of strings (expected: types of child values for a Vec or Map Value)
+// and returns:
+//  - an empty array, iff the input array is empty
+//  - an array with a single element, iff every element within the input array is equal to that element
+//  - the original array, if two or more elements in the input array are different
+// whenever the function returns a singleton array, we can consider the original collection to be homogeneous
+// and its contents to be of the type contained in the singleton return.
+function computeSingleTypeRep(ts: string[]): string[] {
+    if (ts.length === 0) return []
+    else {
+        const t0 = ts[0]
+        const hom = !ts.some((t) => t !== t0)
+        if (hom) return [t0]
+        else return ts
+    }
+}
+
+// Returns the full type annotation of the provided (possibly complex) Value `v`.
+// It uses Apalache-style type notation, with Soroban type primitives.
+// For heterogeneous collections (Vec/Map), the type annotation is distinct, and should be considered to lie
+// outside of the type system.
+export function getFullType(v: Value): string {
+    switch (v.type) {
+        case 'bool':
+        case 'u32':
+        case 'i32':
+        case 'u64':
+        case 'i64':
+        case 'u128':
+        case 'i128':
+        case 'symb':
+        case 'addr':
+            return v.type
+        case 'arr':
+            // As a type, `Bytes` is atomic, even though it is an array. We avoid `arr`, as it could imply
+            // a generic array type, which does not exist.
+            return 'Bytes'
```

Comment:
I do not understand what it means for a type to be atomic. Atomicity means that a value is updated in a single machine instruction, or it is updated in a way such that another thread/process cannot interfere with an update. Do you mean that a value of type `Bytes` is opaque in the sense that it cannot be decomposed into smaller pieces? I am still not sure about the second sentence. What is a generic array type that does not exist?

Reply:
Atomic as in the smallest type-term (or in this context, not a generic collection). E.g. `Array[T]` would be a generic array, which `Bytes` is not. There is `Vec`, which is a dynamic, non-fixed sized collection, but there are no arrays in the Rust sense.

Reply:
I'll rewrite this a bit.